photon-api/src/main/scala/com/linkedin/photon/ml/algorithm/FixedEffectCoordinate.scala

```diff
@@ -14,52 +14,41 @@
  */
 package com.linkedin.photon.ml.algorithm
 
-import org.apache.spark.rdd.RDD
+import org.apache.spark.ml.linalg.{Vector => SparkVector}
+import org.apache.spark.sql.DataFrame
+import org.apache.spark.sql.functions.col
+import org.apache.spark.storage.StorageLevel
 
 import com.linkedin.photon.ml.Types.{FeatureShardId, UniqueSampleId}
+import com.linkedin.photon.ml.constants.DataConst
 import com.linkedin.photon.ml.data._
-import com.linkedin.photon.ml.data.scoring.CoordinateDataScores
 import com.linkedin.photon.ml.function.DistributedObjectiveFunction
 import com.linkedin.photon.ml.model.{DatumScoringModel, FixedEffectModel}
 import com.linkedin.photon.ml.optimization.{DistributedOptimizationProblem, FixedEffectOptimizationTracker, OptimizationTracker}
+import com.linkedin.photon.ml.util.{ApiUtils, VectorUtils}
 
 /**
  * The optimization problem coordinate for a fixed effect model.
  *
  * @tparam Objective The type of objective function used to solve the fixed effect optimization problem
- * @param dataset The training dataset
+ * @param dataset The raw training data
  * @param optimizationProblem The fixed effect optimization problem
+ * @param inputColumnsNames
  */
 protected[ml] class FixedEffectCoordinate[Objective <: DistributedObjectiveFunction](
-    override protected val dataset: FixedEffectDataset,
-    optimizationProblem: DistributedOptimizationProblem[Objective])
-  extends Coordinate[FixedEffectDataset](dataset) {
+    var dataset: DataFrame,
+    optimizationProblem: DistributedOptimizationProblem[Objective],
+    featureShardId: FeatureShardId,
+    inputColumnsNames: InputColumnsNames)
+  extends Coordinate {
```

Comment:
What you described is exactly what is implemented in the comments, but I put the corresponding code logic in different methods (instead of the one you suggested). 

See CoordinateDescent line 192-208: 

      logger.debug(s"Updating coordinate of class ${coordinate.getClass}")

     // compute scores using the previous coordinate model and update offsets
      prevModelOpt.map(model => coordinate.updateOffset(model))
     // Train a new model
      val (model, tracker) = initialModelOpt.map(
        initialModel => Timed(s"Train new model using existing model as starting point") {
          coordinate.trainModel(initialModel)
        }).getOrElse(
        Timed(s"Train new model") {
          coordinate.trainModel()
        })
      // Log summary
      logOptimizationSummary(logger, coordinateId, model, tracker)